pkg/metrics/filters/filters_test.go

```diff
@@ -0,0 +1,253 @@
+package filters
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/prometheus/client_golang/prometheus"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
+)
+
+var _ = Describe("manger.Manager", func() {
+	Describe("Start", func() {
+		Context("should start serving metrics with https and authn/authz", func() {
+			var srv metricsserver.Server
+			var defaultServer metricsDefaultServer
+			var opts manager.Options
+			var httpClient *http.Client
+
+			BeforeEach(func() {
+				srv = nil
+				opts = manager.Options{
+					Metrics: metricsserver.Options{
+						BindAddress:    ":0",
+						SecureServing:  true,
+						FilterProvider: WithAuthenticationAndAuthorization,
+					},
+					NewMetricsServer: func(options metricsserver.Options, config *rest.Config, httpClient *http.Client) (metricsserver.Server, error) {
```

Comment:
I know it's a bit hacky, but I would prefer setting this via reflection in this test instead of exposing the function.

We did something similar here: https://github.com/kubernetes-sigs/controller-runtime/blob/c20ea143a236a34fb331e6c04820b75aac444e7d/pkg/cache/cache_test.go#L546-L553

I think it's just better to have slightly hacky unit tests vs. exposing an option to CR users that we might not want to expose